src/lib/comms/sol-socket-impl-contiki.c

```diff
@@ -0,0 +1,274 @@
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&socket->pending_buffers, buf, idx)
+        free(buf);
+    sol_ptr_vector_clear(&socket->pending_buffers);
+
+    free(socket);
+}
+
+static int
+sol_socket_contiki_set_read_monitor(struct sol_socket *s, bool on)
+{
+    struct sol_socket_contiki *socket = (struct sol_socket_contiki *)s;
+
+    SOL_NULL_CHECK(socket->on_can_read, -EINVAL);
+
+    socket->read_monitor = on;
+
+    return 0;
+}
+
+static bool
+write_timeout_cb(void *data)
+{
+    struct sol_socket_contiki *socket = data;
+
+    if (socket->on_can_write((void *)socket->data, &socket->base))
+        return true;
+
+    socket->write_timeout = NULL;
+    socket->write_monitor = false;
+    return false;
+}
+
+static int
+sol_socket_contiki_set_write_monitor(struct sol_socket *s, bool on)
+{
+    struct sol_socket_contiki *socket = (struct sol_socket_contiki *)s;
+
+    SOL_NULL_CHECK(socket->on_can_write, -EINVAL);
+
+    if (on && !socket->write_timeout) {
+        socket->write_timeout = sol_timeout_add(0, write_timeout_cb, socket);
+        SOL_NULL_CHECK(socket->write_timeout, -ENOMEM);
+    } else if (!on) {
+        sol_timeout_del(socket->write_timeout);
+        socket->write_timeout = NULL;
+    }
+
+    socket->write_monitor = on;
+
+    return 0;
+}
+
+static ssize_t
+sol_socket_contiki_recvmsg(struct sol_socket *s, void *buf, size_t len, struct sol_network_link_addr *cliaddr)
+{
+    struct sol_socket_contiki *socket = (struct sol_socket_contiki *)s;
+    struct pending_buffer *pending_buf;
+    size_t buflen = -EAGAIN;
+
+    pending_buf = sol_ptr_vector_get(&socket->pending_buffers, 0);
+    if (pending_buf) {
+        buflen = pending_buf->datalen;
+        if (buf)
+            sol_ptr_vector_del(&socket->pending_buffers, 0);
+    }
+
+    if (!buf)
+        return buflen;
+
+    if (buflen > len)
+        buflen = len;
+
+    if (cliaddr)
+        *cliaddr = pending_buf->addr;
+
+    memcpy(buf, pending_buf->data, buflen);
+
+    free(pending_buf);
+
+    return buflen;
+}
+
+static ssize_t
+sol_socket_contiki_sendmsg(struct sol_socket *s, const void *buf, size_t len, const struct sol_network_link_addr *cliaddr)
+{
+    struct sol_socket_contiki *socket = (struct sol_socket_contiki *)s;
+
+    simple_udp_sendto_port(&socket->udpconn, buf, len,
+        (uip_ipaddr_t *)&cliaddr->addr.in6, cliaddr->port);
+
+    return len;
+}
+
+static int
+sol_socket_contiki_join_group(struct sol_socket *s, int ifindex, const struct sol_network_link_addr *group)
+{
+    if (group->family != SOL_NETWORK_FAMILY_INET6)
+        return -EAFNOSUPPORT;
+
+    uip_ds6_maddr_add((uip_ipaddr_t *)&group->addr.in6);
+
+    return 0;
+}
+
+static int
+sol_socket_contiki_bind(struct sol_socket *s, const struct sol_network_link_addr *addr)
+{
+    struct sol_socket_contiki *socket = (struct sol_socket_contiki *)s;
+
+    if (addr->family != SOL_NETWORK_FAMILY_INET6)
+        return -EAFNOSUPPORT;
+
+    if (socket->udpconn.udp_conn)
```

Comment:
On Mon, 18 Apr 2016 11:22:21 -0700, Gustavo Lima Chaves wrote:
> > +        return -EAFNOSUPPORT;
> > +
> > +    uip_ds6_maddr_add((uip_ipaddr_t *)&group->addr.in6);
> > +
> > +    return 0;
> > +}
> > +
> > +static int
> > +sol_socket_contiki_bind(struct sol_socket *s, const struct sol_network_link_addr *addr)
> > +{
> > +    struct sol_socket_contiki *socket = (struct sol_socket_contiki *)s;
> > +
> > +    if (addr->family != SOL_NETWORK_FAMILY_INET6)
> > +        return -EAFNOSUPPORT;
> > +
> > +    if (socket->udpconn.udp_conn)
> 
> I guess this is set on simple_udp_register(), right? The patch is fine by me.

Yup, only way I could find to check it hadn't been done already, short
of keeping a variable in the socket.